WalletWasabi/Services/SingleInstanceChecker.cs

```diff
@@ -1,95 +1,117 @@
-		/// <param name="lockName">Name of system-wide mutex.</param>
-		public SingleInstanceChecker(Network network, string lockName)
+		public SingleInstanceChecker(int port)
 		{
-			Network = network;
-			_lockName = $"{MutexString}-{lockName}";
+			Port = port;
 		}
 
-		private IDisposable? SingleApplicationLockHolder { get; set; }
-		private Network Network { get; }
+		private int Port { get; }
+
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private TcpListener? TcpListener { get; set; }
+
+		public event EventHandler? OtherInstanceStarted;
 
 		public async Task CheckAsync()
 		{
-			if (_disposedValue)
+			if (DisposeCts.IsCancellationRequested)
 			{
 				throw new ObjectDisposedException(nameof(SingleInstanceChecker));
 			}
 
-			// The disposal of this mutex handled by AsyncMutex.WaitForAllMutexToCloseAsync().
-			var mutex = new AsyncMutex(_lockName);
 			try
 			{
-				using CancellationTokenSource cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(200));
-				SingleApplicationLockHolder = await mutex.LockAsync(cts.Token).ConfigureAwait(false);
+				TcpListener = new TcpListener(IPAddress.Loopback, Port);
+				TcpListener.Start();
+				await StartAsync(DisposeCts.Token).ConfigureAwait(false);
+
+				// This is the first instance, nothing else to do.
+				return;
 			}
-			catch (IOException ex)
+			catch (SocketException ex) when (ex.ErrorCode is 10048 or 48 or 98)
 			{
-				throw new InvalidOperationException($"Wasabi is already running on {Network}!", ex);
+				// ErrorCodes are different on every OS: win, macOS, lin.
+				// It is already used -> another Wasabi is running on this network.
+				Logger.LogDebug("Detected another Wasabi instance.");
 			}
+
+			// Signal to the other instance, that there was an attempt to start the software.
+			using TcpClient client = new TcpClient();
+			await client.ConnectAsync(IPAddress.Loopback, Port).ConfigureAwait(false);
+			client.Close();
+
+			throw new InvalidOperationException($"Wasabi is already running. Port {Port}!");
 		}
 
-		/// <summary>
-		/// <list type="bullet">
-		/// <item>Unmanaged resources need to be released regardless of the value of the <paramref name="disposing"/> parameter.</item>
-		/// <item>Managed resources need to be released if the value of <paramref name="disposing"/> is <c>true</c>.</item>
-		/// </list>
-		/// </summary>
-		/// <param name="disposing">
-		/// Indicates whether the method call comes from a <see cref="Dispose()"/> method
-		/// (its value is <c>true</c>) or from a finalizer (its value is <c>false</c>).
-		/// </param>
-		protected virtual void Dispose(bool disposing)
+		private static int NetworkToPort(Network network)
 		{
-			if (!_disposedValue)
+			if (network == Network.Main)
 			{
-				if (disposing)
-				{
-					SingleApplicationLockHolder?.Dispose();
-				}
-				_disposedValue = true;
+				return 37129;
 			}
+			else if (network == Network.TestNet)
+			{
+				return 37130;
+			}
+
+			return 37131;
```

Comment:
Segnet network can be added in the future, shouldn't we explicitly check the regtest network?

Reply:
Then we will add it later, ports are valuable resources so if it is not used better to not occupy it.